discovery.go

```diff
@@ -0,0 +1,235 @@
+package pubsub
+
+import (
+	"context"
+	"time"
+
+	"github.com/libp2p/go-libp2p-core/discovery"
+	"github.com/libp2p/go-libp2p-core/peer"
+)
+
+// discover represents the discovery pipeline.
+// The discovery pipeline handles advertising and discovery of peers
+type discover struct {
+	p *PubSub
+
+	// discovery assists in discovering and advertising peers for a topic
+	discovery discovery.Discovery
+
+	// advertising tracks which topics are being advertised
+	advertising map[string]context.CancelFunc
+
+	// discoverQ handles continuing peer discovery
+	discoverQ chan *discoverReq
+
+	// ongoing tracks ongoing discovery requests
+	ongoing map[string]struct{}
+
+	// done handles completion of a discovery request
+	done chan string
+}
+
+type IsDiscovered func(numPeers int) bool
+
+func CheckNumPeers(num int) IsDiscovered {
```

Comment:
What does this do? (needs docs and maybe a better name)

Reply:
refactored, is now `MinTopicSize` option. Current docs for `MinTopicSize`:
```
// MinTopicSize returns a function that checks if a router is ready for publishing based on the topic size.
// The router ultimately decides the whether it is ready or not, the given size is just a suggestion.
```